h/api/storage.py

```diff
@@ -132,24 +165,69 @@ def create_annotation(request, data):
 
     for document_uri_dict in document_uri_dicts:
         models.create_or_update_document_uri(
-            session=request.db,
+            session=session,
             document=document,
             created=annotation.created,
             updated=annotation.updated,
             **document_uri_dict)
 
     for document_meta_dict in document_meta_dicts:
         models.create_or_update_document_meta(
-            session=request.db,
+            session=session,
             document=document,
             created=annotation.created,
             updated=annotation.updated,
             **document_meta_dict)
 
+
+def update_annotation(session, annotation_id, data):
+    """
+    Update an existing annotation and its associated document metadata.
+
+    Update the annotation identified by annotation_id with the given
+    data. Create, delete and update document metadata as appropriate.
+
+    :param session: the database session
+    :type session: sqlalchemy.orm.session.Session
+
+    :param annotation_id: the ID of the annotation to be updated, this is
+        assumed to be a validated ID of an annotation that does already exist
+        in the database
+    :type annotation_id: string
+
+    :param data: the validated data with which to update the annotation
+    :type data: dict
+
+    :returns: the updated annotation
+    :rtype: h.api.models.Annotation
+
+    """
+    # Remove any 'document' field first so that we don't try to save it on the
+    # annotation object.
+    document = data.pop('document', None)
+
+    annotation = models.Annotation.query.get(annotation_id)
+
+    # Just modifying the annotation.extra dict in place doesn't let sqlalchemy
+    # know that it has changed and needs to be saved to the database.
+    # Work around this for now by copying, modifying then assigning.
+    extra = dict(annotation.extra)
+    extra.update(data.pop('extra', {}))
+    annotation.extra = extra
```

Comment:
:+1:
something as the `MutableDict` [here](http://docs.sqlalchemy.org/en/rel_1_0/orm/extensions/mutable.html) should be good, plus a `MutableList` for the `tags` and `references`. Even though we don't mutate some of these data structures but already set the new one on the existing model, I would still like to prevent future bugs by consistently using these special data structures with mutation tracking built-in.
But as you said, that should be done in a separate PR later.